x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/InferModelAction.java

```diff
@@ -0,0 +1,188 @@
+            this.modelVersion = in.readVLong();
+            this.objectsToInfer = Collections.unmodifiableList(in.readList(StreamInput::readMap));
+            this.topClasses = in.readOptionalInt();
+            this.cacheModel = in.readBoolean();
+        }
+
+        public String getModelId() {
+            return modelId;
+        }
+
+        public long getModelVersion() {
+            return modelVersion;
+        }
+
+        public List<Map<String, Object>> getObjectsToInfer() {
+            return objectsToInfer;
+        }
+
+        public boolean isCacheModel() {
+            return cacheModel;
+        }
+
+        public Integer getTopClasses() {
+            return topClasses;
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+            return null;
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            out.writeString(modelId);
+            out.writeVLong(modelVersion);
+            out.writeCollection(objectsToInfer, StreamOutput::writeMap);
+            out.writeOptionalInt(topClasses);
+            out.writeBoolean(cacheModel);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            InferModelAction.Request that = (InferModelAction.Request) o;
+            return Objects.equals(modelId, that.modelId)
+                && Objects.equals(modelVersion, that.modelVersion)
+                && Objects.equals(topClasses, that.topClasses)
+                && Objects.equals(cacheModel, that.cacheModel)
+                && Objects.equals(objectsToInfer, that.objectsToInfer);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(modelId, modelVersion, objectsToInfer, topClasses, cacheModel);
+        }
+
+    }
+
+    public static class RequestBuilder extends ActionRequestBuilder<Request, Response> {
+        public RequestBuilder(ElasticsearchClient client, Request request) {
+            super(client, INSTANCE, request);
+        }
+    }
+
+    public static class Response extends ActionResponse {
+
+        private final List<InferenceResults<?>> inferenceResponse;
+        private final String resultsType;
+
+        public Response(List<InferenceResults<?>> inferenceResponse, String resultsType) {
+            super();
+            this.resultsType = ExceptionsHelper.requireNonNull(resultsType, "resultsType");
+            this.inferenceResponse = inferenceResponse == null ?
+                Collections.emptyList() :
+                Collections.unmodifiableList(inferenceResponse);
+        }
+
+        public Response(StreamInput in) throws IOException {
+            super(in);
+            this.resultsType = in.readString();
+            if(resultsType.equals(ClassificationInferenceResults.RESULT_TYPE)) {
```

Comment:
This polymorphism via `if` here makes me think that `InferenceResults` could be a `NamedWriteable`?

Reply:
It does, but since `InferenceResults<T>` supports a generic, I am not sure it is possible...

I could maybe make `SingleValueInferenceResults` the named writable, but then we would still have a conditional here to choose between single value, multi-value, etc.